### DISCUSSION TITLE
Error simulating using OpenDRIVE standalone mode

### DISCUSSION THREAD

**Discussion by Mayuemei** [carla-simulator/carla#7964] [Q&A]
![stan](https://github.com/user-attachments/assets/23ce0906-92dd-420b-92e1-f231943cf7b8)

I use OpenDRIVE standalone mode（https://carla.readthedocs.io/en/latest/adv_opendrive/） to load a pre written Opendrying file into the reader https://odrviewer.io/ Check if the road can be displayed correctly. Using the following code, it can also be loaded in Carla. Then you want to set up a small car to drive automatically on it, but there may be problems. This problem sometimes occurs and sometimes does not occur. I don't know what the reason is. Here are my Python code and. xodr file, as well as the printed information and error messages. Please help solve it. Thank you very much.

**python代码**

`import random

import carla

import time

def main():
    # Connect to Carla server
    client = carla.Client('localhost', 2000)
    client.set_timeout(10.0)

    # Load custom OpenDrive map
    with open(r'666.xodr', 'r') as od_file:
        data = od_file.read()
    try:
        # Generate the world with the custom OpenDrive map
        world = client.generate_opendrive_world(data)
        print("Custom OpenDrive map loaded successfully.")
    except RuntimeError as e:
        print("Error generating OpenDrive world:", e)
        return

    # Set weather to clear noon for blue sky
    weather = carla.WeatherParameters(
        cloudiness=0.0,
        precipitation=0.0,
        precipitation_deposits=0.0,
        wind_intensity=0.0,
        sun_azimuth_angle=0.0,
        sun_altitude_angle=90.0,
        fog_density=0.0,
        fog_distance=0.0,
        fog_falloff=0.0,
        wetness=0.0,
        scattering_intensity=0.0,
        mie_scattering_scale=0.0,
        rayleigh_scattering_scale=0.0331,
        dust_storm=0.0
    )
    world.set_weather(weather)

    # Clear the world of vehicles and pedestrians
    for actor in world.get_actors():
        if actor.type_id.startswith('vehicle.') or actor.type_id.startswith('walker.'):
            actor.destroy()

    # Get blueprint library
    blueprint_library = world.get_blueprint_library()

    # Select vehicle model
    vehicle_bp = blueprint_library.filter('model3')[0]

    # Set initial spawn point
    spawn_point = carla.Transform(carla.Location(x=10.0, y=0.0, z=0.5), carla.Rotation(pitch=0.0, yaw=0.0, roll=0.0))
    vehicle = world.spawn_actor(vehicle_bp, spawn_point)
    print("Vehicle spawned.")

    # Enable autopilot
    vehicle.set_autopilot(True)
    print("Autopilot enabled.")

    # Get the map object
    map = world.get_map()

    snap = world.wait_for_tick()  # 返回一个仿真快照
    init_frame = snap.frame  # fram代表帧数
    run_frame = 0
    while run_frame < 10000:
        try:
            snap = world.wait_for_tick()
            run_frame = snap.frame - init_frame
            print(run_frame)

            # Set spectator to follow the ego vehicle
            spectator = world.get_spectator()
            ego_tf = vehicle.get_transform()
            spectator_tf = carla.Transform(ego_tf.location, ego_tf.rotation)
            spectator_tf.location += ego_tf.get_forward_vector() * (-10)
            spectator_tf.location += ego_tf.get_up_vector() * 3
            spectator.set_transform(spectator_tf)

            # Get the waypoint at the vehicle's current location
            transform = vehicle.get_transform()
            waypoint = map.get_waypoint(transform.location)
            print(f"-- Current waypoint = {waypoint}")

            # Get current position's lane information
            road_id = waypoint.road_id
            section_id = waypoint.section_id
            lane_id = waypoint.lane_id
            lane_type = waypoint.lane_type
            s = waypoint.s
            print(f" -road_id = {road_id}, section_id = {section_id}, lane_id = {lane_id}, lane_type = {lane_type}, s = {s}")

            # Get nearby waypoints
            left_wp = waypoint.get_left_lane()  # Get waypoints on the left lane
            right_wp = waypoint.get_right_lane()  # Get waypoints on the right lane
            print(f" -left_wp = {left_wp}")
            print(f" -right_wp = {right_wp}")

        except RuntimeError as e:
            print("Error:", e)
            break

        # Destroy vehicle
    is_destroyed = vehicle.destroy()
    if is_destroyed:
        print("Ego vehicle has been destroyed successfully")

if __name__ == '__main__':
    main()
`

**.xodr文件**
`<?xml version="1.0" standalone="yes"?>
<OpenDRIVE>
    <header revMajor="1" revMinor="4" name="" version="1.00" date="2024-07-17" north="0.0" south="0.0" east="0.0" west="0.0">
        <geoReference></geoReference>
    </header>
    <road name="Example Road" length="200" id="1" junction="-1">
        <link/>
        <type s="0.0" type="rural"/>
        <planView>
            <geometry s="0.0" x="0.0" y="0.0" hdg="0.0" length="100.0">
                <line/>
            </geometry>
            <geometry s="100.0" x="100.0" y="0.0" hdg="0.0" length="100.0">
                <arc curvature="0.02"/>
            </geometry>
        </planView>
        <lanes>
            <laneSection s="0.0">
                
                <left>
                    
                    <lane id="2" type="shoulder" level="false">
                        <width sOffset="0.0" a="1.0" b="0.0" c="0.0" d="0.0"/>
                        <roadMark sOffset="0.0000000000000000e+0" type="curb" material="standard" width="0.1" laneChange="none"/>
                    </lane>
                    
                    <lane id="1" type="driving" level="false">
                        <width sOffset="0.0" a="5.0" b="0.0" c="0.0" d="0.0"/>
                        <roadMark sOffset="0.0000000000000000e+0" type="none" material="standard" color="white" laneChange="none"/>
                    </lane>
                </left>
                
                <center>
                    <lane id="0" type="none" level="false"/>
					<roadMark sOffset="0.0000000000000000e+0" type="none" material="standard" color="white" laneChange="none"/>
                </center>
                
                <right>
                    
                    <lane id="-1" type="driving" level="false">
                        <width sOffset="0.0" a="5.0" b="0.0" c="0.0" d="0.0"/>
                        <roadMark sOffset="0.0000000000000000e+0" type="none" material="standard" color="white" laneChange="none"/>
                    </lane>
                    
                    <lane id="-2" type="shoulder" level="false">
                        <width sOffset="0.0" a="1.0" b="0.0" c="0.0" d="0.0"/>
                        <roadMark sOffset="0.0000000000000000e+0" type="curb" material="standard" width="0.1" laneChange="none"/>
                    </lane>
                </right>
            </laneSection>
        </lanes>
    </road>
</OpenDRIVE>`

**Output and error messages：**
Custom OpenDrive map loaded successfully.
Vehicle spawned.
WARNING: cannot parse georeference: ''. Using default values. 
WARNING: No InMemoryMap cache found. Setting up local map. This may take a while... 
Autopilot enabled.
1
-- Current waypoint = Waypoint(Transform(Location(x=10.000259, y=2.500000, z=0.000000), Rotation(pitch=0.000000, yaw=-0.000000, roll=0.000000)))
 -road_id = 1, section_id = 0, lane_id = -1, lane_type = Driving, s = 10.000259399414082
 -left_wp = Waypoint(Transform(Location(x=10.000259, y=-2.500000, z=0.000000), Rotation(pitch=360.000000, yaw=180.000000, roll=0.000000)))
 -right_wp = Waypoint(Transform(Location(x=10.000259, y=5.500000, z=0.000000), Rotation(pitch=0.000000, yaw=-0.000000, roll=0.000000)))
3
Error: trying to operate on a destroyed actor; an actor's function was called, but the actor is already destroyed.
WARNING: cannot parse georeference: ''. Using default values. 
Traceback (most recent call last):
  File "D:/test/test2.py", line 154, in <module>
    main()
  File "D:/test/test2.py", line 147, in main
    is_destroyed = vehicle.destroy()
RuntimeError: trying to operate on a destroyed actor; an actor's function was called, but the actor is already destroyed.

**Comment by Jinviz** [comment]
This problem is usually likely to have a problem with the information in the opendrive file. 
Review the information that defines the opendrive data

**Reply by Mayuemei** [reply]
Thank you for your answer, but it's strange that this issue no longer occurs when I use other control algorithms instead of the autonomous driving in Carla. Sometimes there are indeed errors caused by xodr files. I am currently facing another problem and hope to receive your advice. Thank you very much[https://github.com/carla-simulator/carla/issues/8024](url)
